gadget/install/partition.go

```diff
@@ -167,58 +165,78 @@ func buildPartitionList(dl *gadget.OnDiskVolume, lov *gadget.LaidOutVolume, opts
 
 	// Check if the last partition has a system-data role
 	canExpandData := false
-	if n := len(lov.LaidOutStructure); n > 0 {
-		last := lov.LaidOutStructure[n-1]
-		if last.Role() == gadget.SystemData {
+	if n := len(vol.Structure); n > 0 {
+		last := vol.Structure[n-1]
+		if last.Role == gadget.SystemData {
 			canExpandData = true
 		}
 	}
 
 	// Write new partition data in named-fields format
 	buf := &bytes.Buffer{}
-	for _, laidOut := range lov.LaidOutStructure {
-		if !laidOut.IsPartition() {
+	lastEnd := quantity.Offset(0)
+	toBeCreated = map[int]*gadget.OnDiskStructure{}
+	for _, vs := range vol.Structure {
+		if !vs.IsPartition() {
 			continue
 		}
 
+		// Work out offset, might have not been set if min-size is used
+		// (but note that as we are creating we use the size value)
+		offset := quantity.Offset(0)
+		if vs.Offset != nil {
+			offset = *vs.Offset
+		} else {
+			offset = lastEnd
+		}
+
+		lastEnd = offset + quantity.Offset(vs.Size)
+
 		// Skip partitions defined in the gadget that are already in the volume
-		startInSectors := uint64(laidOut.StartOffset) / sectorSize
+		startInSectors := uint64(offset) / sectorSize
 		if seen[startInSectors] {
 			continue
 		}
 
 		pIndex++
 
 		// Only allow creating certain partitions, namely the ubuntu-* roles
-		if !opts.CreateAllMissingPartitions && !gadget.IsCreatableAtInstall(laidOut.VolumeStructure) {
-			return nil, nil, fmt.Errorf("cannot create partition %s", laidOut)
+		if !opts.CreateAllMissingPartitions && !gadget.IsCreatableAtInstall(&vs) {
+			return nil, nil, fmt.Errorf("cannot create partition #%d (%q)", vs.YamlIndex, vs.Name)
 		}
 
 		// Check if the data partition should be expanded
-		newSizeInSectors := uint64(laidOut.Size) / sectorSize
-		if laidOut.Role() == gadget.SystemData && canExpandData && startInSectors+newSizeInSectors < dl.UsableSectorsEnd {
+		newSizeInSectors := uint64(vs.Size) / sectorSize
+		if vs.Role == gadget.SystemData && canExpandData && startInSectors+newSizeInSectors < dl.UsableSectorsEnd {
 			// note that if startInSectors + newSizeInSectors == dl.UsableSectorEnd
 			// then we won't hit this branch, but it would be redundant anyways
 			newSizeInSectors = dl.UsableSectorsEnd - startInSectors
 		}
 
-		ptype := partitionType(dl.Schema, laidOut.Type())
+		ptype := partitionType(dl.Schema, vs.Type)
 
 		// synthesize the node name and on disk structure
 		node := deviceName(dl.Device, pIndex)
 		// Change bits that depend on the disk, which includes
 		// overriding the size from the gadget as we might be
 		// expanding the data partition.
-		// TODO fill in construction of LaidOutStructure instead if/when possible
-		laidOut.Node = node
-		laidOut.DiskIndex = pIndex
-		laidOut.Size = quantity.Size(newSizeInSectors * sectorSize)
+		vs.Device = node
+		vs.Size = quantity.Size(newSizeInSectors * sectorSize)
```

Comment:
again it's a bit murky to modify the input in place now, and it also gets reflected into the toBeCreated entries